Update Core lightning Application

ShahanaFarooqui opened the pull request:
- Updating environment variables for new application
- Updating image location and environment variable mappings for new application
- Added new application config file
- Updated readme for script location (optional)

nmfretz commented:
Thanks for this @ShahanaFarooqui 🙌. Watching your recent [First Look presentation](https://community.corelightning.org/c/events/check-out-the-cln-umbrel-app) and the app looks awesome! Really great work.

Could I get you to please open a new PR with this app placed under a brand new directory. So for instance, instead of editing the app that is currently in `core-lightning` you would create a new directory called something like `core-lightning-official` or `core-lightning-node'. You can also edit the `umbrel-app.yml` to show that this is the official CLN app.

We will then rename the existing Core Lightning app that was submitted by Umbrel to something like `Simple Core Lightning` to differentiate it from your app which we want to be the main, official CLN app available in the app store.

The reasons for doing this are so that:
1) It is super clear that this is a new official app submitted by CLN.
1) Users who are currently using the umbrel-submitted core lightning app don't feel like they are being forced to switch (although we hope they do).

Let me know if anything is unclear.

rustyrussell commented:
I think we really want to take over "core-lightning": it's the same thing, with a simple GUI on top, and the other applications can then be installed on top, like RTL.

We certainly want to avoid the current case where the user installs core-lightning and it given a screen telling them to install RTL.

lukechilds commented:
Yeah just had a re-think through this internally and we're totally on board with this.

To be clear we weren't planning on continuing both our app and your app, we'd deprecate ours. But after thinking it through it makes more sense to just transition control of our app over to you like you've done in this PR.

The main concern was that we didn't want to hand over control to a different developer in an app update since we'd be essentially handing the user's keys over to a different developer without their knowledge or consent. But that doesn't really make a huge amount of sense in this scenario since anyone using our current Core Lightning app is already trusting you as the maintainers of Core Lightning. Additionally, if we take the approach of launching your app as a new app and deprecating our app then it just causes a load of migration headaches for existing users, potentially encouraging them to stick with a legacy app that won't receive updates which obviously isn't ideal.

Also we have a [`releaseNotes`](https://github.com/getumbrel/umbrel-apps/blob/6babf8eea60d09a2053f8fd7841fffbf9da84a73/core-lightning/umbrel-app.yml#L40) key in the app manifest that is displayed to users as part of the app update UI. We could add some sort of message here like:

> This update brings a stunning new look and feel and is now officially maintained by Blockstream

which makes it clear to the users the app maintainer has changed so they don't feel like things are getting swapped around without their consent.

So if we're proceeding with you guys replacing our app, feel free to update the rest of the parameters in the [app manifest](https://github.com/getumbrel/umbrel-apps/blob/6babf8eea60d09a2053f8fd7841fffbf9da84a73/core-lightning/umbrel-app.yml) to whatever you want.

You can see how the different properties will show up in the app store listing here: 

![](https://camo.githubusercontent.com/bb5dfe1624a6255e12a6ffef2e0e339677f3b53f81a15485e0df731d62438ab1/68747470733a2f2f692e696d6775722e636f6d2f30436f7250524b2e706e67)

nmfretz reviewed (commented):
@ShahanaFarooqui this is looking really good. Some minor comments to address below.

I will test on both a new and existing install. One thing we want to be really sure of is that the app doesn't break for existing users of the Core Lightning app when they update.

As Luke mentioned, please free to update the rest of the parameters in the [app manifest](https://github.com/getumbrel/umbrel-apps/blob/6babf8eea60d09a2053f8fd7841fffbf9da84a73/core-lightning/umbrel-app.yml) to reflect the new app and Blockstream taking over.

We'll also need to update the gallery images/text so that they feature the new UI. Here is what is currently shown https://apps.umbrel.com/app/core-lightning. Do you have anything in mind for the gallery? We can help by working with you to create the images if you want. If you have particular screen shots and text that you would like to use, please pass them along and we can help create these images for you.

ShahanaFarooqui commented:
> We'll also need to update the gallery images/text so that they feature the new UI. Here is what is currently shown https://apps.umbrel.com/app/core-lightning. Do you have anything in mind for the gallery? We can help by working with you to create the images if you want. If you have particular screen shots and text that you would like to use, please pass them along and we can help create these images for you.

I do have screenshots and logo and I will email them to Luke once I am done with other updates. I am currently working on them and hopefully will submit everything by today eod.

ShahanaFarooqui commented on core-lightning/data/app/config.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "unit": "SATS",
+  "fiatUnit": "USD",
+  "appMode": "DARK",
+  "isLoading": false,
+  "error": null
+}
```
Thank you for the suggestion. I will keep it in mind if needed in future updates.

nmfretz replied:
Actually, I'll check if this will even get created correctly when existing users update to the new app the first time... I'm not quite sure.

nmfretz commented:
Thanks Shahana. If you don't mind it's probably best to email them to me directly (nate@umbrel.com) or post them here in this PR.

ShahanaFarooqui commented:
![1](https://user-images.githubusercontent.com/23240721/230232455-926c2f0e-6823-4d2f-ac84-0d3bfe994b33.jpg)

![2](https://user-images.githubusercontent.com/23240721/230232530-09b7f9f0-3f00-40d1-a4a7-23d780b5671c.jpg)

![3](https://user-images.githubusercontent.com/23240721/230232545-2f92854b-09f9-43af-be68-9c0b27e3309d.jpg)

![4](https://user-images.githubusercontent.com/23240721/230232561-15ca0930-6666-4a74-bf8a-aa4c62dc177e.jpg)

ShahanaFarooqui commented:
> Thanks Shahana. If you don't mind it's probably best to email them to me directly ([nate@umbrel.com](mailto:nate@umbrel.com)) or post them here in this PR.

I was unable to paste the svg file here, so sent it in the email to you. Thanks

nmfretz commented:
No worries. Here it is for our reference:

![cln-logo](https://user-images.githubusercontent.com/85373263/230233741-bed3a986-9d48-44d8-a24a-089424e8f2f1.svg)

ShahanaFarooqui commented on core-lightning/docker-compose.yml:
```diff
@@ -7,24 +7,30 @@ services:
+      APP_CORE_LIGHTNING_REST_HIDDEN_SERVICE: http://${APP_CORE_LIGHTNING_REST_HIDDEN_SERVICE}
+      APP_CORE_LIGHTNING_COMMANDO_ENV_DIR: ${APP_CORE_LIGHTNING_COMMANDO_ENV_DIR}
+      APP_CONFIG_DIR: ${APP_CONFIG_DIR}
+      APP_MODE: "production"
```
Thank you for bringing it into my attention. The variable was declared with intention to use but missed somewhere in testing. Using it now.

ShahanaFarooqui commented on core-lightning/docker-compose.yml:
```diff
@@ -7,24 +7,30 @@ services:
     restart: on-failure
     volumes:
-      - ${APP_DATA_DIR}/data/lightningd:/lightningd
+      - ${APP_DATA_DIR}/data/app:/data/app
```
Thank you. Using this and replace some other hard coded values with variable too.

ShahanaFarooqui commented:
@nmfretz Hi Nate. Please review the PR now and let me know if anything else needs to be updated.

nmfretz commented:
@ShahanaFarooqui - Sorry for the delay in getting back to you. I have tested the app on new and existing installs and there are a couple things for us to troubleshoot.

### UI/CLN Connection Issue

The app starts up fine initially after installation; however, after a few minutes the UI becomes unable to connect to CLN and I see this:

<img width="1499" alt="image" src="https://user-images.githubusercontent.com/85373263/232841232-bc5c3f49-f4f4-43dc-aa74-22e05494987f.png">

The UI then becomes accessible again a few minutes later with no intervention. This process of being able to connect and then not being able to connect continues indefinitely. The lightningd container is still up and running during this entire process, but I do see this in the lightningd container logs:
`[ERROR - 5:36:47 PM]: WebSocket is closed`

Is this something you have come across when testing? I wonder if there is some bugginess associated with using 
`--experimental-websocket-port=${APP_CORE_LIGHTNING_WEBSOCKET_PORT}`? 

### Fresh Install vs App Update:

When this app is installed fresh (i.e., with no pre-existing Core Lightning install), the `${APP_DATA_DIR}/data/app` directory and the `config.json` are created with non-root user ownership/permissions (ubuntu in my test case below), which is what we want.

```
~/umbrel/app-data/core-lightning/data$ ls -lR
.:
total 12
drwxrwxr-x 2 ubuntu ubuntu 4096 Apr 18 16:45 app
drwxr-xr-x 3 ubuntu ubuntu 4096 Jan 31 07:01 c-lightning-rest
drwxr-xr-x 3 ubuntu ubuntu 4096 Apr 18 16:45 lightningd

./app:
total 4
-rw-r--r-- 1 root   root     0 Apr 18 16:45 application-cln.log
-rw-rw-r-- 1 ubuntu ubuntu 102 Apr 18 16:44 config.json
```

However, when I update the existing Core Lightning app to the new app, the `${APP_DATA_DIR}/data/app` directory is created with root ownership/permissions (which we do not want) and the `config.json` does not get written to the directory.

```
ls -lR
.:
total 12
drwxr-xr-x 2 root  root  4096 Apr 18 17:11 app
drwxr-xr-x 3 admin admin 4096 Mar 30 17:55 c-lightning-rest
drwxr-xr-x 3 admin admin 4096 Apr  5 22:27 lightningd

./app:
total 4
-rw-r--r-- 1 root root 434 Apr 18 17:10 application-cln.log
```

This results in a status 500 error in the UI when trying to fetch the config on load:
`"ENOENT: no such file or directory, open '/data/app/config.json'"`

If I do something like toggle dark/light mode in the UI, then `config.json` gets written to `${APP_DATA_DIR}/data/app` with root ownership/permissions:

```
./app:
total 8
-rw-r--r-- 1 root root 577 Apr 18 18:03 application-cln.log
-rw-r--r-- 1 root root  84 Apr 18 18:04 config.json
```

I will look further into the dir/file ownership issue.

nmfretz commented:
Here's a first-cut of the gallery images. Would you like to see any changes to the text or images themselves?

![1](https://user-images.githubusercontent.com/85373263/232869485-5e11d1d9-1d4b-4b4e-aeab-e967cdbf2a2f.jpg)

![2](https://user-images.githubusercontent.com/85373263/232869497-d331e62d-b6ad-4a69-8daf-4fbcc0bfe0ce.jpg)

![3](https://user-images.githubusercontent.com/85373263/232869505-5987a83b-5b82-4e13-97af-ecb9f9b9b666.jpg)

ShahanaFarooqui commented:
@nmfretz Hi, I came across this `disconnect and reconnect` issue a while back but that was due to another unrelated issue. It is already fixed by lnmessage library. I have tested this application on my local umbrel setup and it is working as expected. The application is connecting to core lightning node via a websocket and it looks like that is getting closed for some reason. Do you think exposing the websocker port in `lightningd` container can help:
```
    ports:
      - ${APP_CORE_LIGHTNING_DAEMON_PORT}:9735
      - ${APP_CORE_LIGHTNING_WEBSOCKET_PORT}:${APP_CORE_LIGHTNING_WEBSOCKET_PORT}
```

Thank you for looking into the file ownership issue. Please let me know if you want to meet and debug it together.

ShahanaFarooqui commented:
> Here's a first-cut of the gallery images. Would you like to see any changes to the text or images themselves?

Wow, these look beautiful and perfect. Thank you.

nmfretz commented:
> Thank you for looking into the file ownership issue. Please let me know if you want to meet and debug it together.

I think I tracked it down! I'm pretty sure this is due to an issue with our current app update logic for cases where new directories are added. This will be fixed in the next Umbrel release, but we can get around this for now by creating a `pre-start` hook that checks for the existence of the `app` folder and, if it does not exist, creates it and `chown`'s to 1000:1000.

**Would you be okay with me creating this pre-start hook and pushing to your forked repo?**

ShahanaFarooqui commented:
> **Would you be okay with me creating this pre-start hook and pushing to your forked repo?**
That would be great help to me.

nmfretz commented:
> Do you think exposing the websocker port in lightningd container can help:

I'll give this a try!

nmfretz commented:
Okay, I've modified the `pre-start` hook file to:
- create the `${APP_DATA_DIR}/data/app` if it doesn't exist and `chown` to 1000:1000
- create `config.json` with default values if it doesn't exist and `chown` to 1000:1000

This results in the proper dir/file existence and ownership when someone updates from the old app to the new app. The logic will not overwrite an existing config.json on app start/restart/update etc.

@lukechilds - Do you see anything wrong with this approach? As far as I can tell, this is essentially a workaround before this commit from Steve comes out in the next Umbrel release (https://github.com/getumbrel/umbrel/commit/73f99fa7ac930c905cc16d08336edd7dfdd7afcf). 

One downside of this approach, as I've implemented it, is that we now have a template `config.json` in the repo at `/data/app/config.json` as well as a template in the `prestart` hook, so we have duplication that is easy to forget about later.

lukechilds commented:
>This results in a status 500 error in the UI when trying to fetch the config on load:
"ENOENT: no such file or directory, open '/data/app/config.json'"

>If I do something like toggle dark/light mode in the UI, then config.json gets written to ${APP_DATA_DIR}/data/app with root ownership/permissions

Based on the above quotes am I understanding correctly that it's not a permission issue but an issue with the config file not existing when updating from an existing install? The app seems to be able to interact with the volume when it's owned by root because it creates the file when settings are changed.

Your proposed hook looks great Nate. Added some minor suggestions for consistency with other hooks.

lukechilds commented on core-lightning/hooks/pre-start:
```diff
@@ -1,5 +1,33 @@
+
+APP_DATA_DIR="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/..)/data"
+
+[ ! -d "${APP_DATA_DIR}" ] && mkdir -p "${APP_DATA_DIR}" && chown 1000:1000 "${APP_DATA_DIR}"
+
+APP_SUBDIR="${APP_DATA_DIR}/app"
+
+[ ! -d "${APP_SUBDIR}" ] && mkdir -p "${APP_SUBDIR}" && chown 1000:1000 "${APP_SUBDIR}"
```
Prefer `[[` over `[` and `if` instead of `&&` chain.

lukechilds commented:
Also just FYI @ShahanaFarooqui a slightly cleaner way to handle this would be if the CLN UI app auto created the config file if it didn't exist on first start. That would loosen the coupling between your UI and the Umbrel app framework which would be helpful if users want to run this UI without Umbrel. It would also remove the need for duplication here, remove the need for a pre-start hook and give you a bit more control in future app updates if you want to modify the config.

It's not super critical if you're wanting to ship ASAP but i'd say worth looking into for a future update.

lukechilds commented on core-lightning/hooks/pre-start:
```diff
@@ -1,5 +1,33 @@
+
+# Create and set ownership for config.json file if it doesn't exist
+CONFIG_FILE_DESTINATION="${APP_SUBDIR}/config.json"
+CONFIG_FILE_CONTENTS='{
+  "unit": "SATS",
+  "fiatUnit": "USD",
+  "appMode": "DARK",
+  "isLoading": false,
+  "error": null
+}'
```
Re duplication concern, we could technically remove the `data/app/config.json` file I think since this hook should always ensure it's created if it doesn't exist, for both fresh installs and updates.

ShahanaFarooqui commented:
@nmfretz Hi, any update on the PR? Can I help to speed up the process in any way?

ShahanaFarooqui commented:
@nmfretz Hi.

Regarding config file creation: I can include the logic to create the config in runtime if it is missing.

Regarding websocket connection: I did not see that issue happening at my end. Do you have any particular step/suggestion to reproduce it? How much time does it take to connect/disconnect approximately? I always found the application connected whenever I check it at any random time.

nmfretz commented:
> Regarding config file creation: I can include the logic to create the config in runtime if it is missing.

Thanks Shahana. Please feel free to check our logic that we use

> Do you have any particular step/suggestion to reproduce it?

I'll look into this further today. For me personally, the connection issues occur when I keep the UI open in a tab. After a few minutes the UI becomes unable to connect to CLN.

ShahanaFarooqui commented:
> Thanks Shahana. Please feel free to check our logic that we use for the `.conf` files in our Lightning Node and Bitcoin Node apps where we do this.

This is already done and merged in our repo. Will push the change together with below bug fix.

> I'll look into this further today. For me personally, the connection issues occur when I keep the UI open in a tab. After a few minutes the UI becomes unable to connect to CLN.

I was able to reproduce it now by increasing the call frequency. I am currently looking into it.

nmfretz commented:
Also, writing this down here for transparency and so that it's documented for future reference:

> As far as I can tell, this is essentially a workaround before this commit from Steve comes out in the next Umbrel release (https://github.com/getumbrel/umbrel/commit/73f99fa7ac930c905cc16d08336edd7dfdd7afcf).

I was chatting with Luke about this, and I had misunderstood what was happening here. The actual reason for the `config.json` not being created on app **update** is as follows:
- Umbrel only copies **all** the files from the apps repo during install, but not for updates.
- During an app update, Umbrel only copies over a [whitelist of files](https://github.com/getumbrel/umbrel/commit/73f99fa7ac930c905cc16d08336edd7dfdd7afcf#diff-8e674b8e14b37a67c77e07bc9ad10ace548b82c5729fe2c679001b6eedd4dd41R547-R553) from the apps repo, which is intentional because some of the non-whitelisted files could contain state that would then overwrite a user's custom state on update. For example, if `config.json` was copied over from the repo on each update it would overwrite a user's custom settings.
- The volume `${APP_DATA_DIR}/data/app` ends up getting created as root by docker on update, but this doesn't cause any issues in this case, because your app runs as root.

The pre-start hook solves this, but a much more elegant way is to add the logic to the app itself which you have now done.

In the future, our app framework will change to remove this kind of confusion.

ShahanaFarooqui commented:
@nmfretz As discussed in our meeting, below is the PR we need for application auto refresh issue. 

https://github.com/aaronbarnardsound/lnmessage/pull/30

I will update this one as soon as this one is merged and released. Thanks.

nmfretz commented:
Thanks @ShahanaFarooqui. Looking forward to testing and getting this to users!

ShahanaFarooqui commented:
@nmfretz & @lukechilds Both issues mentioned above have been fixed and merged into the PR now. Please check and let me know if anything else needs to be done.

nmfretz commented:
Fantastic work, @ShahanaFarooqui! This is working really well. I have tested on arm64 and amd64 as a fresh install and as an update from the existing Core Lightning app. 

I pushed a commit with the multi-arch digest for the `cln-application` image. The previous digest referred to just the arm image, so the app wouldn't work for users on amd64 devices.

This looks ready to go live in the app store. I would just like @lukechilds to give it a test as well.

nmfretz commented on core-lightning/umbrel-app.yml:
```diff
@@ -8,25 +8,20 @@ description: >-
-  An official app from Umbrel.
-developer: Umbrel
-website: https://umbrel.com
+developer: Shahana Farooqui
```
Are you okay to change this to Core Lightning to match the `submitter` field so that it is super clear to users that this is the official Core Lightning app?

lukechilds commented:
Tested and working great for me, I updated my existing mainnet CLN node and it's looking really slick!

We need to bump the version number in the manifest before merge otherwise users won't get an update notification in the app store: https://github.com/getumbrel/umbrel-apps/blob/7964aaa2c1495c4b8111700a72b31a13b95e78d5/core-lightning/umbrel-app.yml#L5

Also one tiny thing, this is intended to be used to connect external wallets, right?

<img width="729" alt="Screenshot 2023-05-18 at 00 27 46" src="https://github.com/getumbrel/umbrel-apps/assets/2123375/5e8ac435-26e1-413f-a9bf-e49fa160ba97">

If that's the case we'll want to use umbrel.local (`$DEVICE_DOMAIN_NAME`) instead of 10.21.21.96. That IP is a private IP on an internal network and is only reachable from other Umbrel apps. We'll also need to expose `$APP_CORE_LIGHTNING_WEBSOCKET_PORT` on the host.

Other than that, all looking good, excited to see this go out!

ShahanaFarooqui commented:
> We need to bump the version number in the manifest before merge otherwise users won't get an update notification in the app store:

I updated the version to `22.11.2` but the image is still used from version `22.11.1`. Because that is already the latest stable release available for arm64.

> If that's the case we'll want to use umbrel.local ($DEVICE_DOMAIN_NAME) instead of 10.21.21.96. That IP is a private IP on an internal network and is only reachable from other Umbrel apps. 

Replaced IP with domain name.

> We'll also need to expose $APP_CORE_LIGHTNING_WEBSOCKET_PORT on the host.

Exposed the port from `lightningd` service.

Please review.

nmfretz commented:
Thanks Shahana. I have tested on amd64 and arm64 and it is working well.

To keep the version the same as the image-version you can just use something like `22.11.1-ui`. This will still trigger Umbrel to show the user that a Core Lightning update is available, while maintaining the existing versioning structure. Down the road we can think about changing the version to refer to the entire packaged app instead of the clightning version if you wish.

We are ready to launch with your go-ahead! ⚡️⚡️⚡️

nmfretz commented:
We're live! Many thanks to you and your entire team @ShahanaFarooqui for your hard work and patience. This is a wonderful update. https://apps.umbrel.com/app/core-lightning

mayankchhabra commented:
Excellent work on this @ShahanaFarooqui, and huge congrats to the entire Core Lightning team! The new app is sickkkk!

nmfretz commented:
FYI, if you're ever looking to include a badge on your site/readme that shows Core Lightning is available on Umbrel, our [app store preview site](https://apps.umbrel.com/) generates badges for all apps dynamically.


https://apps.umbrel.com/api/app/core-lightning/badge-dark.svg

![](https://apps.umbrel.com/api/app/core-lightning/badge-dark.svg)

https://apps.umbrel.com/api/app/core-lightning/badge-light.svg

![](https://apps.umbrel.com/api/app/core-lightning/badge-light.svg)

https://apps.umbrel.com/badge-dark.svg

![](https://apps.umbrel.com/badge-dark.svg)

https://apps.umbrel.com/badge-light.svg

![](https://apps.umbrel.com/badge-light.svg)


Any/all of them could be simply embedded like:

```html
<a href="https://apps.umbrel.com/app/core-lightning">
  <img src="https://apps.umbrel.com/api/app/core-lightning/badge-dark.svg" />
</a>
```

ShahanaFarooqui commented:
> FYI, if you're ever looking to include a badge on your site/readme that shows Core Lightning is available on Umbrel, our [app store preview site](https://apps.umbrel.com/) generates badges for all apps dynamically.

Thank for sharing the information. Added the badge https://github.com/ElementsProject/cln-application :).

Impa10r commented:
@ShahanaFarooqui hello. The app's web page cannot scroll on Opera (version: 112.0.5197.53) Windows 10 Pro. 

Would be nice to have instructions on how to back up the wallet and a warning before uninstalling it. I saw folks on Reddit losing funds when trying to reinstall.
